File: test/Ch6_SpecialFunctions/factrl/test_factrl.jl
#=
    Unit tests of the factrl function
=#


@testset "test_negative_value" begin
    #=
        Checks values x < 0
    =#
    x = -1
    let err = nothing
        try
            factrl(x)
        catch err
        end
        @test err isa ArgumentError
    end
end


@testset "test_nonInteger_values" begin
    #=
        Checks float, char values
    =#
    x = 1.1
    let err = nothing
        try
            factrl(x)
        catch err
        end
        @test err isa ArgumentError
    end

    x = 'a'
    let err = nothing
        try
            factrl(x)
        catch err
        end
        @test err isa MethodError
    end
end


@testset "test_factrl" begin
    calculated_value = factrl(1)
    expected_value = 1
    @test isapprox(calculated_value, expected_value; atol=1e-8)

    calculated_value = factrl(5)
    expected_value = 120
    @test isapprox(calculated_value, expected_value; atol=1e-8)

    calculated_value = factrl(10)
    expected_value = factorial(10)
    @test isapprox(calculated_value, expected_value; atol=1e-8)

    calculated_value = factrl(15)
    expected_value = factorial(15)
    @test isapprox(calculated_value, expected_value; atol=1e-8)

    calculated_value = factrl(20)
    expected_value = factorial(20)
    @test isapprox(calculated_value, expected_value; atol=1e-8)
end

File: test/Ch6_SpecialFunctions/factrl/test_factrl.py
"""
    Unit tests of the factorial function

    Expected values extracted from Numpy Math
"""
import numericalrecipes as nr
import numpy as np
import unittest as ut


class Test_factrl_input(ut.TestCase):
    """
        Checks values
    """
    def test_negative_value(self):
        """
            Checks values x < 0
        """
        self.assertRaises(ValueError, nr.factrl, -1)

    def test_nonInteger_values(self):
        """
            Checks float, char, etc, values
        """
        self.assertRaises(TypeError, nr.factrl, 1.0)
        self.assertRaises(TypeError, nr.factrl, 'a')
        self.assertRaises(TypeError, nr.factrl, '[0, 1]')

    def test_fctrl(self):
        expected_value = 1
        calculated_value = nr.factrl(1)
        self.assertEqual(calculated_value, expected_value)

        expected_value = 120
        calculated_value = nr.factrl(5)
        self.assertEqual(calculated_value, expected_value)

        expected_value = np.math.factorial(10)
        calculated_value = nr.factrl(10)
        self.assertEqual(calculated_value, expected_value)

        expected_value = np.math.factorial(15)
        calculated_value = nr.factrl(15)
        self.assertEqual(calculated_value, expected_value)

        expected_value = np.math.factorial(20)
        calculated_value = nr.factrl(20)
        self.assertEqual(calculated_value, expected_value)
